File: src/test/java/com/convert/cucumber/steps/ApiSteps.java
package steps;

import io.cucumber.java.en.*;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.response.Response;

import java.util.HashMap;
import java.util.Map;

import static io.restassured.RestAssured.given;
import static org.testng.Assert.*;

public class ApiSteps {

    Response response;
    String tempUser;

    @Given("API server is running")
    public void api_server_is_running() {
        RestAssured.baseURI = "http://127.0.0.1:5000";
    }

    @When("I register via api")
    public void i_register_via_api() {
        tempUser = "user" + System.currentTimeMillis();
        Map<String, String> body = new HashMap<>();
        body.put("username", tempUser);
        body.put("password", "123456");

        response = given()
                .contentType(ContentType.JSON)
                .body(body)
                .post("/api/register");
    }

    @Then("api register response should be success")
    public void api_register_response_should_be_success() {
        assertEquals(response.statusCode(), 200);
        assertEquals(response.jsonPath().getString("msg"), "register success");
    }

    @When("I login via api with username {string} and password {string}")
    public void i_login_via_api(String u, String p) {
        Map<String, String> body = new HashMap<>();
        body.put("username", u);
        body.put("password", p);

        response = given()
                .contentType(ContentType.JSON)
                .body(body)
                .post("/api/login");
    }

    @Then("api login response should be {string}")
    public void api_login_response_should_be(String msg) {
        assertEquals(response.jsonPath().getString("msg"), msg);
    }

    @When("I get user list via api")
    public void i_get_user_list_via_api() {
        response = given().get("/api/users");
    }

    @Then("api user list should not be empty")
    public void api_user_list_should_not_be_empty() {
        assertTrue(response.jsonPath().getList("users").size() > 0);
    }

    @When("I create and delete user via api")
    public void i_create_and_delete_user_via_api() {
        tempUser = "del" + System.currentTimeMillis();
        Map<String, String> body = new HashMap<>();
        body.put("username", tempUser);
        body.put("password", "123456");

        given().contentType(ContentType.JSON).body(body).post("/api/register");
        response = given().get("/api/delete/" + tempUser);
    }

    @Then("api delete response should be success")
    public void api_delete_response_should_be_success() {
        assertEquals(response.jsonPath().getString("msg"), "delete success");
    }
}

File: python_test/behave_demo/features/steps/api_steps.py
from behave import given, when, then
import requests
import time

BASE_URL = "http://127.0.0.1:5000"


@given("API server is running")
def step_api_server(context):
    context.base_url = BASE_URL


@when("I register via api")
def step_register_api(context):
    context.temp_user = f"user{int(time.time())}"
    payload = {
        "username": context.temp_user,
        "password": "123456"
    }
    context.response = requests.post(
        f"{context.base_url}/api/register",
        json=payload
    )


@then("api register response should be success")
def step_register_success(context):
    assert context.response.status_code == 200
    assert context.response.json()["msg"] == "register success"


@when('I login via api with username "{u}" and password "{p}"')
def step_login_api(context, u, p):
    payload = {"username": u, "password": p}
    context.response = requests.post(
        f"{context.base_url}/api/login",
        json=payload
    )


@then('api login response should be "{msg}"')
def step_login_result(context, msg):
    assert context.response.json()["msg"] == msg


@when("I get user list via api")
def step_get_users(context):
    context.response = requests.get(f"{context.base_url}/api/users")


@then("api user list should not be empty")
def step_users_not_empty(context):
    users = context.response.json()["users"]
    assert len(users) > 0


@when("I create and delete user via api")
def step_create_and_delete(context):
    username = f"del{int(time.time())}"
    payload = {"username": username, "password": "123456"}
    requests.post(f"{context.base_url}/api/register", json=payload)
    context.response = requests.get(f"{context.base_url}/api/delete/{username}")


@then("api delete response should be success")
def step_delete_success(context):
    assert context.response.json()["msg"] == "delete success"